File: encryption/encryption.js
//Exports
module.exports = { cipher }


function cipher(batch, keys) {

    //For Each Cycle
    for (let cycleNum = 0; cycleNum < keys.numCycles; cycleNum++) {

        var key = raiseKey(keys.xorKey, cycleNum + 2)
        var keystream = createKeystream(key, batch.length)

        for (let batchIdx = 0; batchIdx < batch.length; batchIdx++) {
            batch[batchIdx] = xor(batch[batchIdx], keystream[batchIdx])
        }
    }

    return batch
}

function xor(num, key) {
    return num ^ key
}

function createKeystream(key, length) {
    var keyStream = Array(length)

    for (let i = 0; i < length; i++) {
        keyStream[i] = key[i % key.length]
    }

    return keyStream
}

function raiseKey(key, num) {
    var raisedKey = Array(key.length)

    for (let i = 0; i < raisedKey.length; i++) {
        raisedKey[i] = (key[i] ** num) % 256
    }

    return raisedKey
}

File: encryption/Encryption.cs
using System;
using System.Collections;
using System.Collections.Generic;
public static class Encryption
{
    public static int[] Cipher(int[] batch, Keys keys)
    {
        for (int cycleNum = 0; cycleNum < keys.numCycles; cycleNum++)
        {
            int[] xorKey = RaiseKey(keys.xorKey, cycleNum + 2);
            int[] keyStream = CreateKeystream(xorKey, batch.Length);

            for (int batchIdx = 0; batchIdx < batch.Length; batchIdx++)
            {
                batch[batchIdx] = XOR(batch[batchIdx], keyStream[batchIdx]);
            }
        }

        return batch;
    }

    static int XOR(int num, int key)
    {
        return num ^ key;
    }

    static int[] CreateKeystream(int[] key, int length)
    {
        int[] keyStream = new int[length];

        for (int i = 0; i < length; i++)
        {
            keyStream[i] = keyStream[i % key.Length];
        }

        return keyStream;
    }

    static int[] RaiseKey(int[] key, int num)
    {
        int[] raisedKey = new int[key.Length];

        for (int i = 0; i < raisedKey.Length; i++)
        {
            raisedKey[i] = Convert.ToInt32(Math.Pow(key[i], num) % 256);
        }

        return raisedKey;
    }
}

public class Keys
{
    public int numCycles;
    public int[] xorKey;
}
